EIPS/eip-4910.md

```diff
@@ -0,0 +1,661 @@
+
+<a name="r55">**[R55]**</a> *The public `royaltyPayOut` function interface MUST adhere to the definition below:*
+```
+/// @dev Function to payout a royalty payment
+/// @param tokenId is the identifier of the NFT token
+/// @param RAsubaccount is the address of the Royalty Sub Account from which the payout should happen
+/// @param receiver is the address to receive the payout
+/// @param amount is the amount to be paid out
+
+function royaltyPayOut (uint256 tokenId, address RAsubaccount, address payable payoutAccount, uint256 amount) public virtual nonReentrant returns (bool)
+```
+The Boolean return value is `true` for a successful function execution, and `false` for an unsuccessful function execution.
+
+Note, that the function has `reentrancy` protection through `nonReentrant` from the Open Zeppelin library since funds are being paid out.
+
+<a name="r56">**[R56]**</a> *The input parameters of the `royaltyPayOut` function MUST satisfy the following requirements:*
+* *`msg.sender == RAsubaccount`.*
+* *`tokenId` must exist and must not be burned.*
+* *`tokenId` must be associated with a Royalty Account.*
+* *`RAsubaccount` must be a valid `accountId` in a Royalty Sub Account of the Royalty Account of the `tokenId'.*
+* *`isIndividual == true` for the Royalty Sub Account, `RAsubaccount`.*
+* *`amount <= royaltybalance` of the Royalty Sub Account, `RAsubaccount.*`
+
+<a name="r57">**[R57]**</a> *The internal `_royaltyPayOut` function interface MUST adhere to the definition below*:
+```
+function _royaltyPayOut (uint256 tokenId, address RAsubaccount, address payable payoutAccount, uint256 amount) public virtual returns (bool)
+```
+
+<a name="r58">**[R58]**</a> *The internal `_royaltyPayOut` function MUST perform the following actions:
+* *send the payment to the `payoutaccount`.*
+* *update the `royaltybalance` of the `RAsubaccount` of the Royalty Account upon successful transfer.*
+
+<a name="r59">**[R59]**</a> *The following steps MUST be taken to send out a royalty payment to its recipient:*
+* *find the Royalty Sub Account.*
+* *extract `tokenType` from the Royalty Sub Account.*
+* *based on the token type send to the `payoutAccount` either*
+    * *'ETH' / relevant protocol token or*
+    * *another token based on token type* 
+* *and only if the payout transaction is successful, deduct `amount` from `royaltybalance` of the Royalty Sub Account,`RAsubaccount`, and then return `true` as the function return parameter, otherwise return `false`.* 
+
+## Backwards Compatibility
+This EIP is backwards compatible to the ERC-721 standard introducing new interfaces and functionality but retaining the core interfaces and functionality of the ERC-721 standard.
+
+## Test Cases
+A full test suite is part of the reference implementation.
+
+## Reference Implementation
+The Treetrunk reference implementation of the standard can be found under https://github.com/ethereum/EIPs/tree/master -> assets -> eip-4910 -> eip-4910-reference-implementation.
```

Comment:
Please link to the implementation using the relative link format `../assets/eip-4910/etc`.

Reply:
@lightclient according to @MicahZoltu I would have to link to every single file in the reference implementation (not feasible due to large number) since folder referencing is not supported in this repo. Please, advise!

Reply:
@MicahZoltu could you elaborate on why a directory listing isn't supported? Is it because the reference implementation should really only be a file or two?

Reply:
Also is an apache 2.0 license okay for a reference implementation?

Reply:
> @MicahZoltu could you elaborate on why a directory listing isn't supported? Is it because the reference implementation should really only be a file or two?

Jekyll (as currently deployed) just doesn't support it.  You'll get a 404.  I'm not totally against adding directory listing support to Jekyll, but we should keep in mind that this would add an additional constraint on any future rendering systems we might want to migrate to in the future.  I generally would rather avoid the constraint and just require people "manually" handle directory listing (e.g., in the EIP include a list of links, or create a single TOC markdown file in assets directory for the EIP in question.

Reply:
> Also is an apache 2.0 license okay for a reference implementation?

No.  Everything in the EIPs repository should be CC0.  If you can't do CC0 (e.g., because it isn't your code), then you should not include it as a reference implementation (just leave out the section).

Reply:
> No. Everything in the EIPs repository should be CC0. If you can't do CC0 (e.g., because it isn't your code), then you should not include it as a reference implementation (just leave out the section).

I'd pretty strongly disagree with this. EIPs themselves should be CC0, but I see no reason to disallow apache/MIT/BSD/Unlicense/etc. They're freely distributable and impose few restrictions on downstream code. Obviously MPL/GPL/AGPL/LGPL/etc would be bad for reference implementations, however.

Reply:
We should probably discuss this among the editors (either in a separate issue or on the fortnightly call).

Reply:
@poojaranjan can we put this on the next agenda?

Reply:
> > @MicahZoltu could you elaborate on why a directory listing isn't supported? Is it because the reference implementation should really only be a file or two?
> 
> Jekyll (as currently deployed) just doesn't support it. You'll get a 404. I'm not totally against adding directory listing support to Jekyll, but we should keep in mind that this would add an additional constraint on any future rendering systems we might want to migrate to in the future. I generally would rather avoid the constraint and just require people "manually" handle directory listing (e.g., in the EIP include a list of links, or create a single TOC markdown file in assets directory for the EIP in question.

@MicahZoltu @lightclient @SamWilsn It might be worth looking into how the Decentralized Identity Foundation, or the W3C, or OASIS handle reference implementations. Generally speaking, reference implementations should come with test fixtures and the ability to generate test reports to show compliance with a standard, at least this is how DIF and W3C handle it. Also, reference implementations are often put into a separate repo that the standard links to. That might be a path forward that could satisfy anyone. Happy to support such an effort.

cc @humbitious

Reply:
If we had more manpower, I would probably agree with improving on the reference implementation constraints.  A big factor in EIP editing decision making is "what requires the least amount of effort" because if effort grows, eventually the whole system grinds to a halt.

Reply:
removed reference implementation